improve: refactor and modularize player forge history handling

dudantas opened the pull request:
### Refactor and modularize player forge history

This PR introduces a new `PlayerForgeHistory` component to encapsulate all forge history logic previously embedded in `Player`.  

#### Key changes
- **New component**  
  - Extracted `ForgeHistory` struct and logic into `PlayerForgeHistory`.  
  - Provides clear `add`, `remove`, `load`, and `save` methods.  
  - Improves encapsulation and separates persistence concerns from `Player`.

- **Database schema and migration**  
  - Added migration `53.lua` to enforce a unique constraint on `done_at` in `forge_history`.  
  - Migration normalizes `done_at` values to milliseconds and resolves duplicates before adding the unique index.  
  - Updated `schema.sql` to include `UNIQUE KEY (done_at)`.

- **Refactored load/save**  
  - `IOLoginDataLoad::loadPlayerForgeHistory` and `IOLoginDataSave::savePlayerForgeHistory` now delegate to the component.  
  - Removed legacy `std::vector<ForgeHistory>` and manual SQL handling from `Player`.  
  - New save logic avoids full table deletes and instead performs targeted deletes and upserts, reducing overhead and improving performance on large datasets.

- **Integration**  
  - `ProtocolGame::sendForgeHistory` and `Player::registerForgeHistoryDescription` updated to use the component.  
  - Cleaned up legacy code paths and redundant SQL queries.

#### Benefits
- Stronger data integrity via enforced uniqueness on `done_at`.  
- Better maintainability by isolating forge history concerns.  
- Performance improvements by eliminating table-wide deletes and using upserts.  
- More robust persistence with clear error logging and benchmarking instrumentation.

dudantas committed:
Refactor and modularize player forge history handling

Moved ForgeHistory struct and related logic from Player to a new PlayerForgeHistory component. Updated database schema and migration to enforce unique 'done_at' in forge_history. Refactored loading and saving of forge history to use the new component, improving encapsulation and maintainability.

dudantas committed:
Merge branch 'main' into dudantas/improve-player-forge-history

dudantas committed:
Fix UNIQUE constraint on done_at in forge_history

dudantas committed:
Refactor PlayerForgeHistory remove and constructor

Replaced manual erase-remove idiom with std::erase_if in PlayerForgeHistory::remove for cleaner code. Marked the PlayerForgeHistory constructor as explicit to prevent implicit conversions.

gabrielew commented on schema.sql:
```diff
@@ -320,7 +320,8 @@ CREATE TABLE IF NOT EXISTS `forge_history` (
     CONSTRAINT `forge_history_pk` PRIMARY KEY (`id`),
-    FOREIGN KEY (`player_id`) REFERENCES `players` (`id`) ON DELETE CASCADE
+    FOREIGN KEY (`player_id`) REFERENCES `players` (`id`) ON DELETE CASCADE,
+    UNIQUE KEY `unique_done_at` (`done_at`)
```
The chances of two players performing a forge action at the exact same time are very low, but it can still happen and may cause an error here, right?
To prevent this, you could add a **unique key** on **player_id** and **done_at**.

```sql
UNIQUE KEY `unique_player_done_at` (`player_id`, `done_at`)
```

dudantas committed:
Change unique key in forge_history table

dudantas committed:
Merge branch 'main' into dudantas/improve-player-forge-history

dudantas committed:
fix: conflicts errors

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Refactor forge history to use player_id and ms timestamps

Updated database migration to enforce uniqueness on (player_id, done_at) instead of just done_at. Refactored PlayerForgeHistory to use uint32_t for IDs, improved duplicate timestamp handling, and switched history creation timestamps to millisecond precision. Adjusted related methods and queries for consistency and correctness.

dudantas committed:
Merge branch 'main' into dudantas/improve-player-forge-history

dudantas committed:
Replace std::any_of with std::ranges::any_of

Updated the add method in PlayerForgeHistory to use std::ranges::any_of for checking duplicate createdAt values, improving code modernity and consistency with C++20 features.
